### PR TITLE
mon: cache reencoded osdmaps

### PR DESCRIPTION
**Pull request by liewegas**
This is a respin of https://github.com/ceph/ceph/pull/21532

http://tracker.ceph.com/issues/23713

### CODE REVIEW

File: src/messages/MOSDMap.h [hunk 1]
```diff
@@ -27,6 +27,7 @@ class MOSDMap : public Message {
 
  public:
   uuid_d fsid;
+  uint64_t encode_features;
```
Comment (tchaikov) [inline src/messages/MOSDMap.h]: please assign a default value to `encode_features`.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-04-25_02:29:48-upgrade:luminous-x-wip-sage2-testing-2018-04-24-1612-distro-basic-smithi/

the crc failures in that run also appear in master, see https://tracker.ceph.com/issues/23871

**Comment by liewegas** [comment]
Rebased and tested (on upgrades from luminous), looking good.  Want to take a final look @tchaikov ?

File: src/mon/OSDMonitor.cc [hunk 2]
```diff
@@ -3481,14 +3499,69 @@ void OSDMonitor::get_removed_snaps_range(
...
+  m.encode(bl, f | CEPH_FEATURE_RESERVED);
+}
+
+int OSDMonitor::get_version(version_t ver, uint64_t features, bufferlist& bl)
+{
+  uint64_t significant_features = OSDMap::get_significant_features(features);
+  if (inc_osd_cache.lookup({ver, significant_features}, &bl)) {
+    return 0;
+  }
+  int ret = PaxosService::get_version(ver, bl);
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: i am reviewing the call sites of `OSDMonitor::get_version(version_t ver, uint64_t features, bufferlist& bl)`. seems they don't like any return value but "0", see https://github.com/ceph/ceph/blob/master/src/mon/OSDMonitor.cc#L378. if the return value is none-zero, `assert(err == 0)` and `assert(!bl.length())`, see https://github.com/ceph/ceph/blob/master/src/mon/OSDMonitor.cc#L377 .

so i am confused, if the callers dislike the `-ENOENT`, why are we take all these efforts to prepare the encoded map for them ? or shall we update the callers or the callee so they are more tolerant in this case? for example, let `OSDMonitor::get_version()` spit a `dout(1)` warning message and returned reencoded bl anyway, but don't return `-ENOENT` to appease the callers.

Comment (xiaoxichen) [reply src/mon/OSDMonitor.cc]: we wont do any reencoding effort in the case of `-NOENT` from paxos service, do we ?

`if (!ret) {`

File: src/mon/OSDMonitor.cc [hunk 3]
```diff
@@ -3481,14 +3499,69 @@ void OSDMonitor::get_removed_snaps_range(
...
+  if (!ret) {
```
Comment (tchaikov) [inline src/mon/OSDMonitor.cc]: i agree. we should *always* reencode the bl retrieved from the underlying monstore. so, https://github.com/ceph/ceph/pull/21605/commits/8b5fb529ed5e19ce3a1cf48335b6219e66da603c#diff-0a5db46a44ae9900e226289a810f10e8R3553 should be 
```c++
if (!ret) {
  return ret;
}
if (siginificant_features != 
    OSDMap::get_significant_features(mon->get_quorum_con_features())) {
  reencode_incremental_map(bl, features);
}
inc_osd_cache.add({ver, significant_features}, bl);
// ...
```

Comment (xiaoxichen) [reply src/mon/OSDMonitor.cc]: It is exactly how it works now,  but yes, your code looks more strightforward

Comment (tchaikov) [reply src/mon/OSDMonitor.cc]: @xiaoxichen please re-read the code. they are different. i am not proposing a cosmetic change.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2018-04-26_17:03:23-rados:thrash-old-clients-wip-sage-testing-2018-04-26-1034-distro-basic-smithi/
also fixed a jewel MOSDMap encoding compat issue with the top patch.

File: src/messages/MOSDMap.h (hunk 1 above)
Comment (liewegas) [reply src/messages/MOSDMap.h]: fixed

File: src/mon/OSDMonitor.cc (hunk 3 above)
Comment (liewegas) [reply src/mon/OSDMonitor.cc]: fixed

File: src/mon/OSDMonitor.cc (hunk 2 above)
Comment (liewegas) [reply src/mon/OSDMonitor.cc]: i think confusion comes from the reencode being inside the if (!ret) block; that's fixed now.

**Comment by liewegas** [comment]
problem with jewel clients still:  https://tracker.ceph.com/issues/23893
<pre>
(gdb) bt
#0  0x00007ff9d0924428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007ff9d092602a in __GI_abort () at abort.c:89
#2  0x00007ff9d126784d in __gnu_cxx::__verbose_terminate_handler() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ff9d12656b6 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ff9d1265701 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ff9d1265919 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ff9d279aa2f in ceph::buffer::list::copy_in (crc_reset=<optimized out>, src=<optimized out>, len=<optimized out>, off=<optimized out>, this=<optimized out>) at common/buffer.cc:1638
#7  0x00007ff9d27a6e8d in ceph::buffer::list::iterator_impl<false>::copy (this=0x7ff9cca4e020, len=<optimized out>, dest=0x7ff9cca4d8c0 "\207��\025�\177") at common/buffer.cc:1125
#8  0x00007ff9d298766b in decode_raw<ceph_le32> (p=..., t=...) at ./include/encoding.h:57
#9  decode (p=..., v=<optimized out>) at ./include/encoding.h:106
#10 decode<int> (p=..., v=...) at ./include/encoding.h:542
#11 decode<pg_t, std::vector<int, std::allocator<int> > > (m=std::map with 1 elements = {...}, p=...) at ./include/encoding.h:648
#12 0x00007ff9d297742d in OSDMap::Incremental::decode (this=this@entry=0x7ff9cca4e160, bl=...) at osd/OSDMap.cc:616
#13 0x00007ff9d282deb2 in OSDMap::Incremental::Incremental (bl=..., this=0x7ff9cca4e160) at ./osd/OSDMap.h:182
#14 Objecter::handle_osd_map (this=this@entry=0x55f832be27a0, m=m@entry=0x7ff9b40008c0) at osdc/Objecter.cc:1146
#15 0x00007ff9d282f0df in Objecter::ms_dispatch (this=0x55f832be27a0, m=0x7ff9b40008c0) at osdc/Objecter.cc:970
#16 0x00007ff9d2a6ae5b in Messenger::ms_deliver_dispatch (m=0x7ff9b40008c0, this=0x55f832be19a0) at ./msg/Messenger.h:584
#17 DispatchQueue::entry (this=0x55f832be1b98) at msg/simple/DispatchQueue.cc:185
#18 0x00007ff9d2a9652d in DispatchQueue::DispatchThread::entry (this=<optimized out>) at msg/simple/DispatchQueue.h:103
#19 0x00007ff9d210c6ba in start_thread (arg=0x7ff9cca50700) at pthread_create.c:333
#20 0x00007ff9d09f641d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
(gdb) f 12
#12 0x00007ff9d297742d in OSDMap::Incremental::decode (this=this@entry=0x7ff9cca4e160, bl=...) at osd/OSDMap.cc:616
</pre>

File: src/mon/OSDMonitor.cc (hunk 3 above)
Comment (tchaikov) [reply src/mon/OSDMonitor.cc]: ahh, sorry. i was wrong. 
```c++
if (!ret) {
  // re-encode
}
return ret;
```
is correct.

Comment (tchaikov) [reply src/mon/OSDMonitor.cc]: sorry, i was wrong. it was me who mis-read the code.

should be 
```c++
if (ret) {
  return ret;
}
// do the re-encoding
return 0;
```

**Pull request merged**